src/tool/cppwinrt/strings/base_deferral.h

```diff
@@ -0,0 +1,72 @@
+namespace winrt
+{
+#ifdef __cpp_coroutines
+    template<typename D>
+    struct deferrable_event_args
+    {
+        Windows::Foundation::Deferral GetDeferral()
+        {
+            slim_lock_guard const guard(m_lock);
+            if (m_handle)
+            {
+                // Cannot ask for deferral after the event handler returned.
+                throw hresult_illegal_method_call();
+            }
+
+            Windows::Foundation::Deferral deferral{ {static_cast<D&>(*this).get_strong(), &deferrable_event_args::one_deferral_completed } };
+            ++m_outstanding_deferrals;
+            return deferral;
+        }
+
+        [[nodiscard]] Windows::Foundation::IAsyncAction wait_for_deferrals()
+        {
+            struct awaitable : std::experimental::suspend_always
+            {
+                bool await_suspend(coroutine_handle handle)
+                {
+                    return m_deferrable.await_suspend(handle);
+                }
+
+                deferrable_event_args& m_deferrable;
+            };
+            co_await awaitable{ {}, *this };
+        }
+
+    private:
+        using coroutine_handle = std::experimental::coroutine_handle<>;
+
+        void one_deferral_completed()
+        {
+            coroutine_handle resume = nullptr;
+            {
+                slim_lock_guard const guard(m_lock);
+                if (m_outstanding_deferrals <= 0)
+                {
+                    throw hresult_illegal_method_call();
+                }
+                if (--m_outstanding_deferrals == 0)
+                {
+                    resume = m_handle;
+                }
+            }
+
+            if (resume)
+            {
+                resume_background().await_suspend(resume);
```

Comment:
Is this too chummy with the implementation of `resume_background()`? Alternatively, this could just be `resume();`, and `wait_for_deferrals()` could be

```cpp
    co_await awaitable{ {}, *this };
    co_await resume_background();
```

Reply:
Since you're now inside the implementation of C++/WinRT I'd say this is acceptable in general, but yes that may be less confusing and perhaps setting a better example for inquisitive developers.

Reply:
I see you're using the `resume_background().await_suspend()` trick in the `await-adapter` branch. Maybe we should factor this out into something less chummy. Move `await_suspend()` to a free function and call it something like `resume_coroutine_on_background_thread()`.

Reply:
Yes, I considered adding a `void winrt::impl::resume_background(coroutine_handle)` helper function to the implementation namespace. We can then keep the name terse and reuse this functionality a bit more broadly.

Reply:
Will leave this open so I can switch to `impl::resume_background()` when it arrives in #546